Backport of:

From ec67af0bd609c412b76eaa4cc89968a2a8e5ad6a Mon Sep 17 00:00:00 2001
From: Jacob Kaplan-Moss <jacob@jacobian.org>
Date: Tue, 13 Aug 2013 11:00:13 -0500
Subject: [PATCH] Fixed is_safe_url() to reject URLs that use a scheme other
 than HTTP/S.

This is a security fix; disclosure to follow shortly.
---
 django/contrib/auth/tests/views.py | 8 ++++++--
 django/utils/http.py               | 7 ++++---
 2 files changed, 10 insertions(+), 5 deletions(-)

Index: python-django-1.3.1/django/contrib/auth/tests/views.py
===================================================================
--- python-django-1.3.1.orig/django/contrib/auth/tests/views.py	2013-09-20 09:16:35.444177420 -0400
+++ python-django-1.3.1/django/contrib/auth/tests/views.py	2013-09-20 09:17:05.344176597 -0400
@@ -261,7 +261,8 @@
         for bad_url in ('http://example.com',
                         'https://example.com',
                         'ftp://exampel.com',
-                        '//example.com'):
+                        '//example.com',
+                        'javascript:alert("XSS")'):
 
             nasty_url = '%(url)s?%(next)s=%(bad_url)s' % {
                 'url': login_url,
@@ -283,6 +284,7 @@
                          '/view?param=ftp://exampel.com',
                          'view/?param=//example.com',
                          'https:///',
+                         'HTTPS:///',
                          '//testserver/',
                          '/url%20with%20spaces/', # see ticket #12534
                          ):
@@ -424,7 +426,8 @@
         for bad_url in ('http://example.com',
                         'https://example.com',
                         'ftp://exampel.com',
-                        '//example.com'
+                        '//example.com',
+                        'javascript:alert("XSS")'
                         ):
             nasty_url = '%(url)s?%(next)s=%(bad_url)s' % {
                 'url': logout_url,
@@ -444,6 +447,7 @@
                          '/view?param=ftp://exampel.com',
                          'view/?param=//example.com',
                          'https:///',
+                         'HTTPS:///',
                          '//testserver/',
                          '/url%20with%20spaces/', # see ticket #12534
                          ):
Index: python-django-1.3.1/django/utils/http.py
===================================================================
--- python-django-1.3.1.orig/django/utils/http.py	2013-09-20 09:16:35.444177420 -0400
+++ python-django-1.3.1/django/utils/http.py	2013-09-20 09:16:35.444177420 -0400
@@ -208,11 +208,12 @@
 def is_safe_url(url, host=None):
     """
     Return ``True`` if the url is a safe redirection (i.e. it doesn't point to
-    a different host).
+    a different host and uses a safe scheme).
 
     Always returns ``False`` on an empty url.
     """
     if not url:
         return False
-    netloc = urlparse.urlparse(url)[1]
-    return not netloc or netloc == host
+    url_info = urlparse.urlparse(url)
+    return (not url_info[1] or url_info[1] == host) and \
+        (not url_info[0] or url_info[0] in ['http', 'https'])
